Tasks/AzureResourceManagerTemplateDeploymentV3/operations/DeploymentScopeBase.ts

```diff
@@ -68,6 +76,58 @@ export class DeploymentScopeBase {
         }
     }
 
+    private printServicePrincipalRoleAssignmentError(error: any) {
+        if(!!error && error.statusCode == 403) {
+            if(this.taskParameters.authScheme == "ServicePrincipal") {
+                if(this.taskParameters.deploymentScope == "Resource Group") {
+                    tl.error(tl.loc("ServicePrincipalRoleAssignmentDetails", this._spnName, this.taskParameters.resourceGroupName));
+                } else if(this.taskParameters.deploymentScope == "Subscription") {
+                    tl.error(tl.loc("ServicePrincipalRoleAssignmentDetails", this._spnName, this.taskParameters.subscriptionId));
+                } else if(this.taskParameters.deploymentScope == "Management Group") {
+                    tl.error(tl.loc("ServicePrincipalRoleAssignmentDetails", this._spnName, this.taskParameters.managementGroupId));    
+                }
+            } else if(this.taskParameters.authScheme == "ManagedServiceIdentity") {
+                tl.error(tl.loc("ManagedServiceIdentityDetails", this.taskParameters.resourceGroupName));    
+            }
+        }
+    }
+
+    protected async getServicePrincipalName(): Promise<string> {
+        try {
+            var graphClient: azureGraph.GraphManagementClient = new azureGraph.GraphManagementClient(this.taskParameters.graphCredentials);
+            var servicePrincipalObject = await graphClient.servicePrincipals.GetServicePrincipal(null);
+            return !!servicePrincipalObject ? servicePrincipalObject.appDisplayName : "";    
+        } catch (error) {
+            tl.debug(tl.loc("ServicePrincipalFetchFailed", error));
+            return "";
+        }
+    }
+
+    protected checkAndPrintPortalDeploymentURL(error: any) {
+        if((this.taskParameters.deploymentScope == "Resource Group" || this.taskParameters.deploymentScope == "Subscription") && (!!error && (error.statusCode < 400 || error.statusCode >= 500))) {
+            tl.error(tl.loc("FindMoreDeploymentDetailsAzurePortal", this.getAzurePortalDeploymentURL()));
+        }
+    }
+
+    private getAzurePortalDeploymentURL() {
+        try {
+            let portalUrl = this.taskParameters.endpointPortalUrl ? this.taskParameters.endpointPortalUrl : "https://portal.azure.com";
+            portalUrl += "/#blade/HubsExtension/DeploymentDetailsBlade/overview/id/";
+    
+            let subscriptionSpecificURL = "/subscriptions/" + this.taskParameters.subscriptionId;
+            if(this.taskParameters.deploymentScope == "Resource Group") {
+                subscriptionSpecificURL += "/resourceGroups/" + this.taskParameters.resourceGroupName;
+            }
+    
+            subscriptionSpecificURL += "/providers/Microsoft.Resources/deployments/" + this.taskParameters.deploymentName;
+    
+            return portalUrl + subscriptionSpecificURL.replace(/\//g, '%2F');
+        } catch (error) {
+            tl.error(error);
+            return error;
```

Comment:
why return the error? This will get printed as deployment URL which is not good. May be print the `FindMoreDeploymentDetailsAzurePortal` only of this function returns a non-empty message.

Reply:
Done